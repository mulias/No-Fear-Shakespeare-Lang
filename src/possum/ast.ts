export type Program = { type: "program"; value: (Node | Malformed)[] };

export type Node = Operand | Prefix | Infix | Postfix;

export type Operand = Number | Character | Var | Comment | DocComment;

export type Prefix = {
  type: "negate";
  prefixed: Node;
};

export type Infix = {
  type: "method_access" | "modulo" | "multiply" | "divide" | "add" | "subtract";
  left: Node;
  right: Node;
};

export type Postfix = {
  type: "function_call" | "block";
  value: (Node | Malformed)[];
  postfixed: Node;
};

export type Number = Integer | InvalidNumber;

export type Integer = { type: "int"; value: number };

export type InvalidNumber = { type: "invalid_number"; value: number };

export type Character = { type: "char"; value: string };

export type Var = { type: "var"; value: string };

export type Comment = { type: "comment"; value: string };

export type DocComment = {
  type: "doc_comment";
  value: [key: string, value: string];
};

export type Malformed = { type: "malformed"; value: string };
