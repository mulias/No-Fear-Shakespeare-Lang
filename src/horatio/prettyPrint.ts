import * as Ast from "./ast";
import Formatter, { Part, Subpart, Line } from "./formatter";

export const prettyPrint = (ast: Ast.Program): string => {
  const formatter = new Formatter();
  const { title, declarations, parts } = ast.visit(formatter);

  return [title, declarations.join("\n"), ...parts.map(prettyPrintPart)].join(
    "\n\n",
  );
};

const prettyPrintPart = ({ heading, subparts }: Part): string => {
  return [indent(heading, 20), ...subparts.map(prettyPrintSubpart)].join(
    "\n\n",
  );
};

const prettyPrintSubpart = ({ heading, body }: Subpart): string => {
  return [indent(heading, 20), ...body.map(prettyPrintLine)].join("\n\n");
};

const prettyPrintLine = (line: Line | string): string => {
  if (typeof line === "string") {
    return line;
  } else {
    return [line.name, indent(wrap(line.text, 76), 4)].join("\n");
  }
};

const indent = (text: string, spaces: number): string => {
  const indentation = " ".repeat(spaces);

  return text
    .split("\n")
    .map((line) => indentation + line)
    .join("\n");
};

const wrap = (text: string, maxLength: number): string => {
  const words = text.split(" ");
  let lines: string[] = [];
  let line = "";

  words.forEach((word) => {
    if (line.length + word.length + 1 < maxLength) {
      line = line + " " + word;
    } else {
      lines.push(line);
      line = word;
    }
  });

  lines.push(line);

  return lines.join("\n");
};
