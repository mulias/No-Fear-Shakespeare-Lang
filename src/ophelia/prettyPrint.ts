import * as Ast from "./ast";
import { AssertNeverError } from "../util";

export function prettyPrint(ast: Ast.Program): string {
  const printer = new PrettyPrinter();
  return printer.printProgram(ast);
}

class PrettyPrinter {
  private indentLevel = 0;
  private readonly indentSize = 2;

  private indent(): string {
    return " ".repeat(this.indentLevel * this.indentSize);
  }

  private withIndent<T>(fn: () => T): T {
    this.indentLevel++;
    const result = fn();
    this.indentLevel--;
    return result;
  }

  private printTemplateString(templateString: Ast.TemplateString): string {
    return templateString.value
      .map((segment) => {
        if (segment.type === "template_string_segment") {
          return segment.value;
        } else if (segment.type === "template_var_segment") {
          return `{${segment.value}}`;
        } else {
          throw new Error(
            `Unknown template segment type: ${(segment as any).type}`,
          );
        }
      })
      .join("");
  }

  printProgram(program: Ast.Program): string {
    const parts: string[] = [];

    if (program.title) {
      parts.push(`## title: ${this.printTemplateString(program.title)}`);
    }

    // Add var declarations
    for (const [varName, description] of program.varDeclarations) {
      parts.push(`## var ${varName}: ${this.printTemplateString(description)}`);
    }

    if (program.items.length > 0) {
      // Add empty line before first act if there are var declarations or title
      if (program.title || program.varDeclarations.size > 0) {
        parts.push("");
      }

      // Print each program item
      for (let i = 0; i < program.items.length; i++) {
        const programItem = program.items[i];
        if (programItem) {
          const printedItem = this.printProgramItem(programItem);
          parts.push(printedItem);

          // Add blank line after acts (except the last one)
          if (programItem.type === "act" && i < program.items.length - 1) {
            parts.push("");
          }
        }
      }
    }

    return parts.join("\n");
  }

  private printProgramItem(item: Ast.ProgramItem): string {
    switch (item.type) {
      case "act":
        return this.printAct(item);
      case "comment":
        return this.printComment(item);
      default:
        throw new AssertNeverError(item);
    }
  }

  private printAct(act: Ast.Act): string {
    const parts: string[] = [];

    // Add description doc comment if present
    if (act.description) {
      parts.push(
        `## description: ${this.printTemplateString(act.description)}`,
      );
    }

    const header = `${act.actId} {`;
    const footer = "}";

    if (act.items.length === 0) {
      parts.push(`${header}\n${footer}`);
      return parts.join("\n");
    }

    const items = this.withIndent(() => {
      const itemParts: string[] = [];
      for (let i = 0; i < act.items.length; i++) {
        const item = act.items[i];
        if (item) {
          itemParts.push(this.printActItem(item));

          if (i < act.items.length - 1) {
            itemParts.push("");
          }
        }
      }
      return itemParts.join("\n");
    });

    parts.push(`${header}\n${items}\n${footer}`);
    return parts.join("\n");
  }

  private printActItem(item: Ast.ActItem): string {
    switch (item.type) {
      case "scene":
        return this.printScene(item);
      case "comment":
        return this.printComment(item);
      default:
        throw new AssertNeverError(item);
    }
  }

  private printScene(scene: Ast.Scene): string {
    const parts: string[] = [];

    // Add description doc comment if present
    if (scene.description) {
      parts.push(
        `${this.indent()}## description: ${this.printTemplateString(
          scene.description,
        )}`,
      );
    }

    const header = `${this.indent()}${scene.sceneId} {`;
    const footer = `${this.indent()}}`;

    if (scene.directions.length === 0) {
      parts.push(`${header}\n${footer}`);
      return parts.join("\n");
    }

    const directions = this.withIndent(() => {
      const dirParts: string[] = [];
      for (let i = 0; i < scene.directions.length; i++) {
        const dir = scene.directions[i];
        if (dir) {
          dirParts.push(this.printDirection(dir));

          // Apply spacing rules based on current and next direction types
          if (i < scene.directions.length - 1) {
            const nextDir = scene.directions[i + 1];
            if (nextDir) {
              const currentIsStage = this.isStageDirection(dir);
              const nextIsStage = this.isStageDirection(nextDir);

              // Add a blank line unless both items are stage directions, in
              // which case they should be grouped
              if (!currentIsStage || !nextIsStage) {
                dirParts.push("");
              }
            }
          }
        }
      }
      return dirParts.join("\n");
    });

    parts.push(`${header}\n${directions}\n${footer}`);
    return parts.join("\n");
  }

  private isStageDirection(direction: Ast.Direction): boolean {
    return (
      direction.type === "stage" ||
      direction.type === "unstage" ||
      direction.type === "unstage_all"
    );
  }

  private printDirection(direction: Ast.Direction): string {
    switch (direction.type) {
      case "dialogue":
        return this.printDialogue(direction);
      case "stage":
        return this.printStage(direction);
      case "unstage":
        return this.printUnstage(direction);
      case "unstage_all":
        return `${this.indent()}unstage_all`;
      case "comment":
        return this.printComment(direction);
      default:
        throw new AssertNeverError(direction);
    }
  }

  private printComment(comment: Ast.Comment): string {
    return `${this.indent()}# ${comment.content}`;
  }

  private printDialogue(dialogue: Ast.Dialogue): string {
    const header = `${this.indent()}${dialogue.speakerVarId} {`;
    const footer = `${this.indent()}}`;

    if (dialogue.lines.length === 0) {
      return `${header}\n${footer}`;
    }

    const lines = this.withIndent(() => {
      const lineParts: string[] = [];
      for (let i = 0; i < dialogue.lines.length; i++) {
        const line = dialogue.lines[i];
        if (line) {
          lineParts.push(this.printStatementOrComment(line));

          // Add blank line after statement if it has followedByBlankLine flag
          // and it's not the last statement in the dialogue
          if (i < dialogue.lines.length - 1) {
            if (line.type !== "comment" && line.followedByBlankLine) {
              lineParts.push("");
            }
          }
        }
      }
      return lineParts.join("\n");
    });

    return `${header}\n${lines}\n${footer}`;
  }

  private printStatementOrComment(item: Ast.StatementOrComment): string {
    if (item.type === "comment") {
      return this.printComment(item);
    } else {
      return this.printStatement(item);
    }
  }

  private printStage(stage: Ast.Stage): string {
    const args = stage.varIds.join(", ");
    return `${this.indent()}stage(${args})`;
  }

  private printUnstage(unstage: Ast.Unstage): string {
    const args = unstage.varIds.join(", ");
    return `${this.indent()}unstage(${args})`;
  }

  private printStatement(statement: Ast.Statement): string {
    switch (statement.type) {
      case ".set":
        return `${this.indent()}@you.set(${this.printExpression(
          statement.value,
        )})`;
      case ".print_char":
        return `${this.indent()}@you.print_char`;
      case ".print_int":
        return `${this.indent()}@you.print_int`;
      case ".read_char":
        return `${this.indent()}@you.read_char`;
      case ".read_int":
        return `${this.indent()}@you.read_int`;
      case ".push_self":
        return `${this.indent()}@you.push_self`;
      case ".push_me":
        return `${this.indent()}@you.push_me`;
      case ".pop":
        return `${this.indent()}@you.pop`;
      case "test_eq":
      case "test_gt":
      case "test_lt":
      case "test_not_eq":
      case "test_not_gt":
      case "test_not_lt":
        return `${this.indent()}${statement.type}(${this.printExpression(
          statement.left,
        )}, ${this.printExpression(statement.right)})`;
      case "if":
        const condition = statement.is ? "if_true" : "if_false";
        return `${this.indent()}${condition}(${this.printStatement(
          statement.then,
        ).trim()})`;
      case "goto":
        return `${this.indent()}goto(${statement.labelId})`;
      default:
        throw new AssertNeverError(statement);
    }
  }

  private printExpression(expression: Ast.Expression): string {
    return this.printExpressionWithPrecedence(expression, 0);
  }

  private printExpressionWithPrecedence(
    expression: Ast.Expression,
    parentPrecedence: number,
  ): string {
    switch (expression.type) {
      case "int":
        return expression.value.toString();
      case "char":
        return `'${expression.value}'`;
      case "var":
        return expression.id;
      case "you":
        return "@you";
      case "arithmetic":
        const precedence = this.getOperatorPrecedence(expression.op);
        const left = this.printExpressionWithPrecedence(
          expression.left,
          precedence,
        );
        const right = this.printExpressionWithPrecedence(
          expression.right,
          precedence,
        );
        const result = `${left} ${expression.op} ${right}`;

        // Add parentheses if this operator has lower precedence than parent
        if (precedence < parentPrecedence) {
          return `(${result})`;
        }
        return result;
      case "unary":
        return `${expression.op}(${this.printExpressionWithPrecedence(
          expression.operand,
          0,
        )})`;
      default:
        throw new AssertNeverError(expression);
    }
  }

  private getOperatorPrecedence(op: Ast.ArithmeticOp): number {
    // PEMDAS precedence: higher number = higher precedence
    switch (op) {
      case "*":
      case "/":
      case "%":
        return 2; // Multiplication, Division, Modulo
      case "+":
      case "-":
        return 1; // Addition, Subtraction
      default:
        return 0;
    }
  }
}
