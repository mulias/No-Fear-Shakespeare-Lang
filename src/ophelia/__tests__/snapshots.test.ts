import { Possum } from "../../possum";
import { Ophelia } from "../index";
import { prettyPrint } from "../prettyPrint";

describe("NFSPL Round-trip Snapshots", () => {
  /**
   * Parse NFSPL source and pretty print it back to NFSPL
   */
  async function roundTripNfspl(source: string): Promise<string> {
    // Step 1: Parse NFSPL with Possum
    const possum = new Possum(source);
    const possumAst = await possum.run();

    // Step 2: Transform with Ophelia
    const ophelia = new Ophelia(possumAst);
    const opheliaAst = ophelia.run();

    // Step 3: Pretty print back to NFSPL
    return prettyPrint(opheliaAst);
  }

  describe("basic programs", () => {
    it("should round-trip a hello world program", async () => {
      const source = `Main {
  PrintHi {
    stage(a, b)

    b {
      @you.set(72)
      @you.print_char
    }

    b {
      @you.set(@you + 1)
      @you.print_char
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip a program with multiple scenes", async () => {
      const source = `Main {
  Start {
    stage(n, out)

    out {
      @you.set(0)
    }
  }

  Loop {
    out {
      @you.set(@you + 1)
      test_gt(@you, 100)
      if_true(goto(End))
    }
  }

  End {
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });
  });

  describe("language features", () => {
    it("should round-trip I/O operations", async () => {
      const source = `Main {
  IOTest {
    stage(reader, writer)

    reader {
      @you.read_char
      @you.read_int
    }

    writer {
      @you.set('A')
      @you.print_char
      @you.set(42)
      @you.print_int
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip arithmetic operations", async () => {
      const source = `Main {
  ArithmeticTest {
    stage(calc, result)

    calc {
      @you.set(10 + 5)
      @you.set(20 - 7)
      @you.set(6 * 7)
      @you.set(15 / 3)
      @you.set(17 % 5)
    }

    result {
      @you.set(calc + result)
      @you.set(@you * 2)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip stack operations", async () => {
      const source = `Main {
  StackTest {
    stage(pusher, stack)

    pusher {
      @you.set(42)
      @you.push_self
      @you.push_me
    }

    stack {
      @you.pop
      @you.pop
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip control flow", async () => {
      const source = `Main {
  ControlTest {
    stage(tester, value)

    tester {
      test_eq(@you, 42)
      if_true(goto(Success))
      test_gt(@you, 0)
      if_false(goto(Failure))
    }
  }

  Success {
    stage(winner, output)

    winner {
      @you.set('W')
      @you.print_char
    }

    unstage_all
  }

  Failure {
    stage(loser, output)

    loser {
      @you.set('L')
      @you.print_char
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip all comparison operations", async () => {
      const source = `Main {
  ComparisonTest {
    stage(left, right)

    left {
      test_eq(@you, right)
      test_gt(@you, right)
      test_lt(@you, right)
      test_not_eq(@you, right)
      test_not_gt(@you, right)
      test_not_lt(@you, right)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip character literals", async () => {
      const source = `Main {
  CharTest {
    stage(char_writer, output)

    char_writer {
      @you.set('A')
      @you.set('Z')
      @you.set('0')
      @you.set('9')
      @you.set(' ')
      @you.set('!')
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip stage and unstage operations", async () => {
      const source = `Main {
  StageTest {
    stage(a)
    unstage(a)
    stage(b, c)
    unstage(b, c)
    stage(d, e)
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip nested arithmetic expressions", async () => {
      const source = `Main {
  NestedMath {
    stage(calc, temp)

    calc {
      @you.set((5 + 3) * 2)
      @you.set(@you / 4 - 1)
      @you.set((temp + @you) % 10)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip if statements with different actions", async () => {
      const source = `Main {
  IfTest {
    stage(controller, output)

    controller {
      test_eq(@you, 0)
      if_true(@you.print_char)
      if_false(@you.print_int)
      if_true(@you.push_self)
      if_false(@you.pop)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip goto statements in various contexts", async () => {
      const source = `Main {
  GotoTest {
    stage(jumper, counter)

    jumper {
      goto(NextScene)
    }
  }

  NextScene {
    stage(controller, value)

    controller {
      test_eq(@you, 0)
      if_true(goto(End))
      goto(GotoTest)
    }
  }

  End {
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip multiple acts", async () => {
      const source = `Act1 {
  Scene1 {
    stage(a, b)

    a {
      @you.set(1)
    }

    unstage_all
  }

  Scene2 {
    stage(c, d)

    c {
      @you.set(2)
    }

    unstage_all
  }
}

Act2 {
  Scene1 {
    stage(e, f)

    e {
      @you.set(3)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip multiple dialogues in one scene", async () => {
      const source = `Main {
  Dialogue {
    stage(a, b)

    a {
      @you.set('H')
      @you.print_char
    }

    b {
      @you.set('i')
      @you.print_char
    }

    a {
      @you.set('!')
      @you.print_char
    }

    b {
      @you.set(10)
      @you.print_char
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip mixed @you operations", async () => {
      const source = `Main {
  YouTest {
    stage(speaker, listener)

    speaker {
      @you.set(@you)
      @you.set(@you + 1)
      test_eq(@you, listener)
      test_gt(listener, @you)
    }

    listener {
      @you.set(speaker + @you)
      @you.push_self
      @you.push_me
      @you.pop
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });
  });

  describe("complex programs", () => {
    it("should round-trip fizzbuzz structure", async () => {
      const source = `Main {
  Start {
    stage(n, out)

    out {
      @you.set(0)
    }
  }

  Loop {
    out {
      @you.set(@you + 1)
    }

    out {
      test_gt(@you, 100)
      if_true(goto(End))
    }

    out {
      test_eq(@you % 15, 0)
      if_true(goto(Fizzbuzz))
    }

    out {
      test_eq(@you % 3, 0)
      if_true(goto(Fizz))
    }

    out {
      test_eq(@you % 5, 0)
      if_true(goto(Buzz))
    }

    n {
      @you.set(out)
      @you.print_int
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Fizzbuzz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Fizz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Buzz {
    n {
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  End {
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip reverse program structure", async () => {
      const source = `Main {
  Init {
    stage(stack, count)

    stack {
      @you.set(0)
    }
  }

  GetInput {
    count {
      @you.read_char
      @you.push_self
    }

    stack {
      @you.set(@you + 1)
      test_gt(@you, 0)
    }

    count {
      if_true(goto(GetInput))
      @you.pop
    }
  }

  PrintReversed {
    count {
      @you.pop
      @you.print_char
    }

    stack {
      @you.set(@you - 1)
      test_gt(@you, 0)
    }

    count {
      if_true(goto(PrintReversed))
    }
  }

  End {
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });
  });

  describe("formatting fixes", () => {
    it("should fix inconsistent indentation", async () => {
      const source = `Main{
Start{
stage(a,b)

   a{
@you.set(72)
      @you.print_char
}

        b{
    @you.set(@you+1)
  @you.print_char
          }

unstage_all
}
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should fix missing spaces around operators", async () => {
      const source = `Main {
  Math {
    stage(calc,temp)

    calc {
      @you.set(5+3*2)
      @you.set(@you/4-1)
      @you.set((temp+@you)%10)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should fix spacing in function calls", async () => {
      const source = `Main {
  TestSpacing {
    stage(tester,value)

    tester {
      test_eq(@you,42)
      if_true(goto(End))
      test_gt( @you , 0 )
      if_false( goto(Start) )
    }
  }

  End {
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should fix spacing in stage/unstage calls", async () => {
      const source = `Main {
  StageSpacing {
    stage(a,b)
    unstage( a , b )
    stage( c )
    unstage(c)
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });
  });

  describe("edge cases", () => {
    it("should round-trip empty scenes", async () => {
      const source = `Main {
  EmptyScene {
  }

  AnotherEmpty {
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip scenes with only stage directions", async () => {
      const source = `Main {
  StageOnly {
    stage(a, b)
    unstage(a)
    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip minimal dialogue blocks", async () => {
      const source = `Main {
  MinimalDialogue {
    stage(speaker)

    speaker {
      @you.set(42)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });

    it("should round-trip single character variables", async () => {
      const source = `Main {
  SingleChar {
    stage(a, b)

    a {
      @you.set(b)
    }

    b {
      @you.set(a + 1)
    }

    unstage_all
  }
}`;

      const result = await roundTripNfspl(source);
      expect(result).toMatchSnapshot();
    });
  });
});
