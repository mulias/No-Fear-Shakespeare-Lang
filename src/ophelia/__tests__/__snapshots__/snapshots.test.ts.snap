// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFSPL Round-trip Snapshots basic programs should round-trip a hello world program 1`] = `
"Main {
  PrintHi {
    stage(a, b)

    b {
      @you.set(72)
      @you.print_char
    }

    b {
      @you.set(@you + 1)
      @you.print_char
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots basic programs should round-trip a program with multiple scenes 1`] = `
"Main {
  Start {
    stage(n, out)

    out {
      @you.set(0)
    }
  }

  Loop {
    out {
      @you.set(@you + 1)
      test_gt(@you, 100)
      if_true(goto(End))
    }
  }

  End {
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots complex programs should round-trip fizzbuzz structure 1`] = `
"Main {
  Start {
    stage(n, out)

    out {
      @you.set(0)
    }
  }

  Loop {
    out {
      @you.set(@you + 1)
    }

    out {
      test_gt(@you, 100)
      if_true(goto(End))
    }

    out {
      test_eq(@you % 15, 0)
      if_true(goto(Fizzbuzz))
    }

    out {
      test_eq(@you % 3, 0)
      if_true(goto(Fizz))
    }

    out {
      test_eq(@you % 5, 0)
      if_true(goto(Buzz))
    }

    n {
      @you.set(out)
      @you.print_int
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Fizzbuzz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Fizz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Buzz {
    n {
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  End {
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots complex programs should round-trip reverse program structure 1`] = `
"Main {
  Init {
    stage(stack, count)

    stack {
      @you.set(0)
    }
  }

  GetInput {
    count {
      @you.read_char
      @you.push_self
    }

    stack {
      @you.set(@you + 1)
      test_gt(@you, 0)
    }

    count {
      if_true(goto(GetInput))
      @you.pop
    }
  }

  PrintReversed {
    count {
      @you.pop
      @you.print_char
    }

    stack {
      @you.set(@you - 1)
      test_gt(@you, 0)
    }

    count {
      if_true(goto(PrintReversed))
    }
  }

  End {
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots edge cases should round-trip empty scenes 1`] = `
"Main {
  EmptyScene {
  }

  AnotherEmpty {
  }
}"
`;

exports[`NFSPL Round-trip Snapshots edge cases should round-trip minimal dialogue blocks 1`] = `
"Main {
  MinimalDialogue {
    stage(speaker)

    speaker {
      @you.set(42)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots edge cases should round-trip scenes with only stage directions 1`] = `
"Main {
  StageOnly {
    stage(a, b)
    unstage(a)
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots edge cases should round-trip single character variables 1`] = `
"Main {
  SingleChar {
    stage(a, b)

    a {
      @you.set(b)
    }

    b {
      @you.set(a + 1)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots formatting fixes should fix inconsistent indentation 1`] = `
"Main {
  Start {
    stage(a, b)

    a {
      @you.set(72)
      @you.print_char
    }

    b {
      @you.set(@you + 1)
      @you.print_char
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots formatting fixes should fix missing spaces around operators 1`] = `
"Main {
  Math {
    stage(calc, temp)

    calc {
      @you.set(5 + 3 * 2)
      @you.set(@you / 4 - 1)
      @you.set((temp + @you) % 10)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots formatting fixes should fix spacing in function calls 1`] = `
"Main {
  TestSpacing {
    stage(tester, value)

    tester {
      test_eq(@you, 42)
      if_true(goto(End))
      test_gt(@you, 0)
      if_false(goto(Start))
    }
  }

  End {
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots formatting fixes should fix spacing in stage/unstage calls 1`] = `
"Main {
  StageSpacing {
    stage(a, b)
    unstage(a, b)
    stage(c)
    unstage(c)
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip I/O operations 1`] = `
"Main {
  IOTest {
    stage(reader, writer)

    reader {
      @you.read_char
      @you.read_int
    }

    writer {
      @you.set('A')
      @you.print_char
      @you.set(42)
      @you.print_int
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip all comparison operations 1`] = `
"Main {
  ComparisonTest {
    stage(left, right)

    left {
      test_eq(@you, right)
      test_gt(@you, right)
      test_lt(@you, right)
      test_not_eq(@you, right)
      test_not_gt(@you, right)
      test_not_lt(@you, right)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip arithmetic operations 1`] = `
"Main {
  ArithmeticTest {
    stage(calc, result)

    calc {
      @you.set(10 + 5)
      @you.set(20 - 7)
      @you.set(6 * 7)
      @you.set(15 / 3)
      @you.set(17 % 5)
    }

    result {
      @you.set(calc + result)
      @you.set(@you * 2)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip character literals 1`] = `
"Main {
  CharTest {
    stage(char_writer, output)

    char_writer {
      @you.set('A')
      @you.set('Z')
      @you.set('0')
      @you.set('9')
      @you.set(' ')
      @you.set('!')
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip control flow 1`] = `
"Main {
  ControlTest {
    stage(tester, value)

    tester {
      test_eq(@you, 42)
      if_true(goto(Success))
      test_gt(@you, 0)
      if_false(goto(Failure))
    }
  }

  Success {
    stage(winner, output)

    winner {
      @you.set('W')
      @you.print_char
    }

    unstage_all
  }

  Failure {
    stage(loser, output)

    loser {
      @you.set('L')
      @you.print_char
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip goto statements in various contexts 1`] = `
"Main {
  GotoTest {
    stage(jumper, counter)

    jumper {
      goto(NextScene)
    }
  }

  NextScene {
    stage(controller, value)

    controller {
      test_eq(@you, 0)
      if_true(goto(End))
      goto(GotoTest)
    }
  }

  End {
    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip if statements with different actions 1`] = `
"Main {
  IfTest {
    stage(controller, output)

    controller {
      test_eq(@you, 0)
      if_true(@you.print_char)
      if_false(@you.print_int)
      if_true(@you.push_self)
      if_false(@you.pop)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip mixed @you operations 1`] = `
"Main {
  YouTest {
    stage(speaker, listener)

    speaker {
      @you.set(@you)
      @you.set(@you + 1)
      test_eq(@you, listener)
      test_gt(listener, @you)
    }

    listener {
      @you.set(speaker + @you)
      @you.push_self
      @you.push_me
      @you.pop
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip multiple acts 1`] = `
"Act1 {
  Scene1 {
    stage(a, b)

    a {
      @you.set(1)
    }

    unstage_all
  }

  Scene2 {
    stage(c, d)

    c {
      @you.set(2)
    }

    unstage_all
  }
}

Act2 {
  Scene1 {
    stage(e, f)

    e {
      @you.set(3)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip multiple dialogues in one scene 1`] = `
"Main {
  Dialogue {
    stage(a, b)

    a {
      @you.set('H')
      @you.print_char
    }

    b {
      @you.set('i')
      @you.print_char
    }

    a {
      @you.set('!')
      @you.print_char
    }

    b {
      @you.set(10)
      @you.print_char
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip nested arithmetic expressions 1`] = `
"Main {
  NestedMath {
    stage(calc, temp)

    calc {
      @you.set((5 + 3) * 2)
      @you.set(@you / 4 - 1)
      @you.set((temp + @you) % 10)
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip stack operations 1`] = `
"Main {
  StackTest {
    stage(pusher, stack)

    pusher {
      @you.set(42)
      @you.push_self
      @you.push_me
    }

    stack {
      @you.pop
      @you.pop
    }

    unstage_all
  }
}"
`;

exports[`NFSPL Round-trip Snapshots language features should round-trip stage and unstage operations 1`] = `
"Main {
  StageTest {
    stage(a)
    unstage(a)
    stage(b, c)
    unstage(b, c)
    stage(d, e)
    unstage_all
  }
}"
`;
